pkg/api/plugin.go

```diff
@@ -14,6 +14,34 @@ const (
 	ContextKeyTenantID     ContextKey = "TenantID"
 )
 
+type PluginStep string
+
+const (
+	PluginStepDeploy                  PluginStep = "Deploy"
+	PluginStepInitialize              PluginStep = "InitializeCluster"
+	PluginStepUpdateDrain             PluginStep = "UpdateDrain"
+	PluginStepUpdateMasterVMRotation  PluginStep = "UpdateMasterVMRotation"
+	PluginStepUpdateInfraVMRotation   PluginStep = "UpdateInfraVMRotation"
+	PluginStepUpdateComputeVMRotation PluginStep = "UpdateComputeVMRotation"
+	PluginStepWaitForHealthz          PluginStep = "WaitForHealthz"
+	PluginStepWaitForNodes            PluginStep = "WaitForNodes"
+	PluginStepHealthCheck             PluginStep = "HealthCheck"
+	PluginStepWaitForInfraDaemonSets  PluginStep = "WaitForInfraDaemonSets"
+	PluginStepWaitForInfraDeployments PluginStep = "WaitForInfraDeployments"
+)
+
+// PluginError error returned by CreateOrUpdate to specify the step that failed.
+type PluginError struct {
```

Comment:
Thoughts on making this an interface and introducing  functions to ensure that a `PluginError` is always well formed?  Suggestions:

* error cannot be nil
* step cannot be empty
* constructor could return nil if error is nil and panic if step is empty (so that the last call of methods could be `return NewPluginError(err, step)`)

Reply:
I think the idea is that "error" is the interface that we use. My first time making extended errors, so I have not done the right thing here. I should be returning "error" everywhere and making sure to explicitly return "nil" where possible.

Reply:
things can implement [multiple interfaces](https://play.golang.org/p/IQ60H3YN41q).  Shouldn't be an issue returning `error` in signatures, even from a `NewPluginError` constructor.  

I guess I should note that Kube just uses a struct for their structured validation errors (in apimachinery utils).  But it does seem like we want to police the structure so we do not end up with uncategorized errors or a `PluginError` with a nil error.  If folk agree it can be done with code review - just want to make sure we're deciding.

Reply:
For the record, I'm fine with this as-is.  Most of the time we want to check nil errors anyway unless we're at the end of a method